# Setup and Load Data

In [96]:
import numpy as np
import pandas as pd

In [97]:
def read_n_create_index(cityname):
    data = pd.read_csv("rent_listings_"+cityname+".csv")
    data['city_id'] = cityname + '_' +  data.index.astype(str)
    data.set_index('city_id', inplace=True)
    print("Size of the data:", data.shape)
    display(data.head())
    return data

In [98]:
lagos_data = read_n_create_index('lagos')

Size of the data: (5000, 6)


neighborhood        price  \
city_id                                                                   
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  ₦ 2,500,000   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  ₦ 2,500,000   
lagos_2                             -Old Ikoyi Ikoyi Lagos     $ 40,000   
lagos_3                       ...Lekki Phase 1 Lekki Lagos    ₦ 800,000   
lagos_4                          Lekki Phase 1 Lekki Lagos  ₦ 1,200,000   

           beds    baths    toilets  \
city_id                               
lagos_0  3 beds  4 baths  4 Toilets   
lagos_1  3 beds  3 baths  4 Toilets   
lagos_2  3 beds  3 baths  4 Toilets   
lagos_3  1 beds  1 baths  1 Toilets   
lagos_4  1 beds  1 baths  1 Toilets   

                                                     extra  
city_id                                                     
lagos_0  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_2  Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_3  For more information, Pricing, and Private Vie...  
lagos_4  For more information, Pricing, and Private Vie...

In [99]:
abuja_data = read_n_create_index('abuja')

Size of the data: (739, 6)


neighborhood        price    beds  \
city_id                                                                         
abuja_0       Sam najuma estate GaladimawaLokogoma Abuja  ₦ 1,300,000  3 beds   
abuja_1                             DawakiGwarinpa Abuja  ₦ 1,500,000  3 beds   
abuja_2                                     Gaduwa Abuja  ₦ 1,500,000  3 beds   
abuja_3                                    Maitama Abuja  ₦ 2,500,000  2 beds   
abuja_4  In an estate behind games village Garki 1 Abuja  ₦ 1,500,000  2 beds   

           baths    toilets                                              extra  
city_id                                                                         
abuja_0  3 baths  4 Toilets  Sharp 2nd floor 3 bedroom apartment @ Dr Sam n...  
abuja_1  3 baths  4 Toilets  Newly BuiltIt's newly built and never been occ...  
abuja_2  3 baths  4 Toilets  3bedroom blocks of flat in Gaduwa District, gr...  
abuja_3  2 baths  4 Toilets  ServicedA Serviced 2 bedroom flat in #maitama ...  
abuja_4  2 baths  3 Toilets  2 Bedroom Flat With CCTV, Constant Water Suppl...

In [100]:
ogun_data = read_n_create_index('ogun')

Size of the data: (188, 6)


neighborhood      price    beds    baths  \
city_id                                                                         
ogun_0             ShimawaMowe Obafemi Owode Ogun  ₦ 260,000  3 beds  1 baths   
ogun_1                       Adigbe Abeokuta Ogun  ₦ 300,000  2 beds  3 baths   
ogun_2   Private EstateMagboro Obafemi Owode Ogun  ₦ 250,000  1 beds  1 baths   
ogun_3          MagboroMagboro Obafemi Owode Ogun  ₦ 350,000  2 beds  2 baths   
ogun_4                   Private EstateArepo Ogun  ₦ 450,000  2 beds  2 baths   

           toilets                                              extra  
city_id                                                                
ogun_0   2 Toilets  Newly BuiltA newly built 1 bed, 2 beds and 3 b...  
ogun_1   3 Toilets  2 bedroom Shared Apartment Flat / Apartment fo...  
ogun_2   2 Toilets  Spacious Rooms Ample Parking Space Kitchen Cab...  
ogun_3   2 Toilets  24-hours Security Spacious Rooms Ample Parking...  
ogun_4   3 Toilets  24-hours Security Spacious Rooms Ample Parking...

In [101]:
ibadan_data = read_n_create_index('oyo')

Size of the data: (512, 6)


neighborhood        price  \
city_id                                                                   
oyo_0     Iyaganku GRA quarters IbadanIyanganku Ibadan Oyo  ₦ 1,000,000   
oyo_1    Behind DSS ile tuntun nihort Jericho ibadanIdi...    ₦ 450,000   
oyo_2     Yidi apete junction ibadan Ibadan Polytechnic...    ₦ 126,000   
oyo_3    Adeoyo GRA, off ring road, ibadan Ring Rd Ibad...    ₦ 750,000   
oyo_4    Opposition nnpc filling station, akala express...    ₦ 600,000   

           beds    baths    toilets  \
city_id                               
oyo_0    3 beds  3 baths  3 Toilets   
oyo_1    3 beds  3 baths  3 Toilets   
oyo_2    1 beds  1 baths  1 Toilets   
oyo_3    3 beds  3 baths  4 Toilets   
oyo_4    3 beds  3 baths  4 Toilets   

                                                     extra  
city_id                                                     
oyo_0    *To let Three bedroom flat en suit is availabl...  
oyo_1    *To let* 4 flat of 3 bedroom with 3 toilets, b...  
oyo_2    Serviced Newly BuiltA room selfcon at Yidi ape...  
oyo_3    Newly BuiltFinely finished 3 bedroom Apartment...  
oyo_4    Newly BuiltBrand new 3 bedroom Apartment for r...

In [102]:
ph_data = read_n_create_index('rivers')

Size of the data: (105, 6)


neighborhood        price  \
city_id                                                                    
rivers_0  GRA phase 3 tombia extensionNew Gra Port Harco...  ₦ 1,500,000   
rivers_1                    Ada George Port Harcourt Rivers    ₦ 500,000   
rivers_2         odili roadTrans Amadi Port Harcourt Rivers    ₦ 550,000   
rivers_3                   Trans Amadi Port Harcourt Rivers    ₦ 650,000   
rivers_4                ...Trans Amadi Port Harcourt Rivers    ₦ 800,000   

            beds    baths    toilets  \
city_id                                
rivers_0  2 beds  2 baths  2 Toilets   
rivers_1  2 beds  2 baths  2 Toilets   
rivers_2  1 beds  1 baths  2 Toilets   
rivers_3    beds    baths    Toilets   
rivers_4  0 beds  0 baths  0 Toilets   

                                                      extra  
city_id                                                      
rivers_0  Newly Built2no newly built 2bed room flats in ...  
rivers_1  Furnished Newly BuiltA VIRGIN TWO BEDROOM FLAT...  
rivers_2  One bedroom for [email protected] #550k odili ...  
rivers_3  Decent King Size 2 Bedroom Flat For Rent In Wo...  
rivers_4  King Size 2 Bedroom For Rent In Woji  Port Har...

## Data Lookup and Cleaning

In [103]:
data = pd.concat([lagos_data, abuja_data, ibadan_data, ogun_data, ph_data])
print("Size of the data:", data.shape)

data.head()

Size of the data: (6544, 6)


neighborhood        price  \
city_id                                                                   
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  ₦ 2,500,000   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  ₦ 2,500,000   
lagos_2                             -Old Ikoyi Ikoyi Lagos     $ 40,000   
lagos_3                       ...Lekki Phase 1 Lekki Lagos    ₦ 800,000   
lagos_4                          Lekki Phase 1 Lekki Lagos  ₦ 1,200,000   

           beds    baths    toilets  \
city_id                               
lagos_0  3 beds  4 baths  4 Toilets   
lagos_1  3 beds  3 baths  4 Toilets   
lagos_2  3 beds  3 baths  4 Toilets   
lagos_3  1 beds  1 baths  1 Toilets   
lagos_4  1 beds  1 baths  1 Toilets   

                                                     extra  
city_id                                                     
lagos_0  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_2  Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_3  For more information, Pricing, and Private Vie...  
lagos_4  For more information, Pricing, and Private Vie...

### Cleaning Price;

In [104]:
data.price.unique()

array(['₦ 2,500,000', '$ 40,000', '₦ 800,000', '₦ 1,200,000',
       '₦ 1,000,000', '₦ 350,000', '₦ 550,000', '₦ 700,000',
       '₦ 2,000,000', '₦ 500,000', '₦ 400,000', '₦ 600,000', '₦ 300,000',
       '₦ 900,000', '₦ 1,500,000', '₦ 1,800,000', '₦ 850,000',
       '₦ 170,000', '₦ 1,600,000', '₦ 650,000', '₦ 750,000',
       '₦ 1,300,000', '₦ 1,100,000', '₦ 2,300,000', '₦ 250,000',
       '₦ 1,400,000', '₦ 450,000', '₦ 200,000', '₦ 120,000',
       '₦ 1,700,000', '₦ 950,000', '₦ 55,000', '₦ 130,000', '₦ 1,900,000',
       '₦ 65,000', '₦ 1,850,000', '₦ 150,000', '₦ 1,650,000', '₦ 470,000',
       '₦ 1,350,000', '₦ 270,000', '₦ 35,000', '$ 25,000', '$ 65,000',
       '₦ 2,200,000', '$ 100,000', '₦ 2,250,000', '$ 80,000', '$ 28,000',
       '₦ 70,000', '$ 50,000', '₦ 2,100,000', '₦ 840,000', '$ 70,000',
       '₦ 210,000', '₦ 216,000', '₦ 250', '₦ 84,000', '₦ 160,000',
       '₦ 1,025,000', '₦ 230,000', '₦ 280,000', '₦ 480,000', '₦ 430,000',
       '₦ 220,000', '₦ 2,400,000', '₦ 330,000'

Some of the prices are in dollars! We'd probably drop those rows.

In [105]:
dollar_apartments = data[data.price.str.startswith('$')]
dollar_apartments

neighborhood      price  \
city_id                                                                    
lagos_2                                -Old Ikoyi Ikoyi Lagos   $ 40,000   
lagos_398                               Victoria Island Lagos   $ 25,000   
lagos_399                               Victoria Island Lagos   $ 65,000   
lagos_404                               Old Ikoyi Ikoyi Lagos  $ 100,000   
lagos_408                               Old Ikoyi Ikoyi Lagos   $ 80,000   
lagos_412                               Victoria Island Lagos   $ 28,000   
lagos_433                               Old Ikoyi Ikoyi Lagos   $ 50,000   
lagos_605                                         Ikoyi Lagos   $ 70,000   
lagos_1689     Ozumba MbadiweBonny Camp Victoria Island Lagos   $ 75,000   
lagos_1874                -Eko Atlantic Victoria Island Lagos   $ 50,000   
lagos_2268              Ligali Ayorinde Victoria Island Lagos   $ 80,000   
lagos_2294  victoria island Victoria Island Victoria Islan...   $ 65,000   
lagos_2351  YOUNIS BASHORUNVictoria Island Extension Victo...   $ 80,000   
lagos_2504            Banana Island Banana Island Ikoyi Lagos   $ 90,000   
lagos_2572    Adeola Odeku Kofo Abayomi Victoria Island Lagos   $ 25,000   
lagos_2581                 Eko Atlantic Victoria Island Lagos   $ 90,000   
lagos_2800             Banana IslandBanana Island Ikoyi Lagos   $ 90,000   
lagos_3130  Kofo Abayomi Ahmadu Bello Way Victoria Island ...   $ 12,000   
lagos_4082                 Eko Atlantic Victoria Island Lagos   $ 70,000   
lagos_4893             Banana IslandBanana Island Ikoyi Lagos   $ 65,000   
lagos_4979               VIAkin Adesola Victoria Island Lagos   $ 80,000   
abuja_75                        Oyi river streetMaitama Abuja   $ 35,000   
abuja_81                        Oyi river StreetMaitama Abuja   $ 35,000   
abuja_124                          Osun crescentMaitama Abuja   $ 18,000   
abuja_140                               Maitama Maitama Abuja   $ 68,000   
abuja_143                                MaitamaMaitama Abuja   $ 25,000   
abuja_292                          Sunrise HillsAsokoro Abuja   $ 65,000   
abuja_305                          Sunrise HillsAsokoro Abuja   $ 65,000   
abuja_334                         Maitama, AbujaMaitama Abuja  $ 105,000   
abuja_546                               Asokoro Asokoro Abuja   $ 45,000   

              beds    baths     toilets  \
city_id                                   
lagos_2     3 beds  3 baths   4 Toilets   
lagos_398   3 beds  3 baths   4 Toilets   
lagos_399   3 beds  3 baths   4 Toilets   
lagos_404   3 beds  3 baths   4 Toilets   
lagos_408   3 beds  3 baths   4 Toilets   
lagos_412   1 beds  1 baths   2 Toilets   
lagos_433   3 beds  3 baths   4 Toilets   
lagos_605   3 beds  3 baths   4 Toilets   
lagos_1689  3 beds  3 baths   4 Toilets   
lagos_1874  3 beds  3 baths   4 Toilets   
lagos_2268  3 beds  4 baths   4 Toilets   
lagos_2294  3 beds  0 baths   0 Toilets   
lagos_2351  3 beds  3 baths   4 Toilets   
lagos_2504  3 beds  3 baths   4 Toilets   
lagos_2572  1 beds  2 baths   2 Toilets   
lagos_2581  3 beds  3 baths   4 Toilets   
lagos_2800  3 beds  3 baths   4 Toilets   
lagos_3130  1 beds  1 baths   1 Toilets   
lagos_4082  4 beds  4 baths   5 Toilets   
lagos_4893  4 beds  4 baths   5 Toilets   
lagos_4979  3 beds  3 baths   4 Toilets   
abuja_75    2 beds  2 baths   3 Toilets   
abuja_81    2 beds  2 baths   3 Toilets   
abuja_124   3 beds  3 baths   4 Toilets   
abuja_140   3 beds  3 baths   4 Toilets   
abuja_143   3 beds  3 baths   4 Toilets   
abuja_292   3 beds  3 baths   4 Toilets   
abuja_305   3 beds  3 baths   4 Toilets   
abuja_334   4 beds  4 baths   5 Toilets   
abuja_546   3 beds  3 baths   4 Toilets   

                                                        extra  
city_id                                                        
lagos_2     Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_398   Modern spacious flat in Victoria Is

They all seems to be located at either Lagos or Abuja.

In [106]:
data.drop(dollar_apartments.index, inplace=True)

data.price.unique()

array(['₦ 2,500,000', '₦ 800,000', '₦ 1,200,000', '₦ 1,000,000',
       '₦ 350,000', '₦ 550,000', '₦ 700,000', '₦ 2,000,000', '₦ 500,000',
       '₦ 400,000', '₦ 600,000', '₦ 300,000', '₦ 900,000', '₦ 1,500,000',
       '₦ 1,800,000', '₦ 850,000', '₦ 170,000', '₦ 1,600,000',
       '₦ 650,000', '₦ 750,000', '₦ 1,300,000', '₦ 1,100,000',
       '₦ 2,300,000', '₦ 250,000', '₦ 1,400,000', '₦ 450,000',
       '₦ 200,000', '₦ 120,000', '₦ 1,700,000', '₦ 950,000', '₦ 55,000',
       '₦ 130,000', '₦ 1,900,000', '₦ 65,000', '₦ 1,850,000', '₦ 150,000',
       '₦ 1,650,000', '₦ 470,000', '₦ 1,350,000', '₦ 270,000', '₦ 35,000',
       '₦ 2,200,000', '₦ 2,250,000', '₦ 70,000', '₦ 2,100,000',
       '₦ 840,000', '₦ 210,000', '₦ 216,000', '₦ 250', '₦ 84,000',
       '₦ 160,000', '₦ 1,025,000', '₦ 230,000', '₦ 280,000', '₦ 480,000',
       '₦ 430,000', '₦ 220,000', '₦ 2,400,000', '₦ 330,000', '₦ 180,000',
       '₦ 530,000', '₦ 10,000', '₦ 380,000', '₦ 420,000', '₦ 90,000',
       '₦ 110,000', '₦ 500

No dollar-priced apartments anymore!

In [107]:
data.price = data.price.str.replace(',', '').apply(lambda x: x[2:])

data.price.unique()

array(['2500000', '800000', '1200000', '1000000', '350000', '550000',
       '700000', '2000000', '500000', '400000', '600000', '300000',
       '900000', '1500000', '1800000', '850000', '170000', '1600000',
       '650000', '750000', '1300000', '1100000', '2300000', '250000',
       '1400000', '450000', '200000', '120000', '1700000', '950000',
       '55000', '130000', '1900000', '65000', '1850000', '150000',
       '1650000', '470000', '1350000', '270000', '35000', '2200000',
       '2250000', '70000', '2100000', '840000', '210000', '216000', '250',
       '84000', '160000', '1025000', '230000', '280000', '480000',
       '430000', '220000', '2400000', '330000', '180000', '530000',
       '10000', '380000', '420000', '90000', '110000', '500', '1150000',
       '140000', '1250000', '25000', '100000', '96000', '40000', '45000',
       '460000', '630000', '72000', '370000', '6', '620000', '1200009',
       '50000', '320000', '1750000', '60000', '270', '150', '350',
       '190000', '520

There are instances where the price value is less than 4 digits. Having a closer look;

In [108]:
cheap_apartments = data[data.price.str.len() <= 5]
cheap_apartments

neighborhood  price    beds  \
city_id                                                                        
lagos_178                           Lekki Phase 1 Lekki Lagos  55000  3 beds   
lagos_204                             ...Abule Oja Yaba Lagos  65000  1 beds   
lagos_397   Shonibare Estate Mobolaji Bank Anthony Way Ike...  35000  1 beds   
lagos_430                         Parkview Estate Ikoyi Lagos  70000  3 beds   
lagos_802   Madonna Estate ojodu Abiodun road Berger Ojodu...    250  1 beds   
lagos_820    Unity street,unity bus stopIgbogbo Ikorodu Lagos  84000  1 beds   
lagos_1344                                   Akoka Yaba Lagos  10000    beds   
lagos_1696  Itele ogun state close to ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_1753                           Ojodu Berger Ojodu Lagos    500  3 beds   
lagos_1866                              Phase 2 Gbagada Lagos  25000  1 beds   
lagos_2071  Nasfat street before Agunfoye sawmill along ig...  96000  1 beds   
lagos_2136  Off freedom road lekki phase 1Lekki Phase 1 Le...  40000  1 beds   
lagos_2213  Off studio 24, Lekki phase 1 Lekki Phase 1 Lek...  45000  2 beds   
lagos_2346  Itele Ogun state close to Ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_2363  Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
lagos_2493             Rumens Road IkoyiOld Ikoyi Ikoyi Lagos      6  1 beds   
lagos_2502                              Igbogbo Ikorodu Lagos  90000  1 beds   
lagos_2543                Off Lekki Epe Expressway Ajah Lagos  70000  2 beds   
lagos_2615  Very decent and beautiful new house 2bedroom a...  50000  2 beds   
lagos_2653  Itele Ogun State close to Ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_2697                         Spar roadIkate Lekki Lagos  45000  1 beds   
lagos_2839  Asekun street off adekunle kuyeAdelabu Suruler...    500  1 beds   
lagos_3112  Off Admirathy road Lekki phase 1 Lekki Phase 1...  25000  1 beds   
lagos_3143                  Itele ayobo roadAyobo Ipaja Lagos  60000  1 beds   
lagos_3194  Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
lagos_3208  Okotie Eboh street off Awolowo road ikoyiAwolo...  40000  1 beds   
lagos_3222                 College road Ifako Ogba Ogba Lagos    270    beds   
lagos_3225        3 Greenville estate badoreBadore Ajah Lagos    150  1 beds   
lagos_3241          Abule-ijesha road Abule Ijesha Yaba Lagos    350    beds   
lagos_3320                          Lekki Phase 1 Lekki Lagos  50000  3 beds   
lagos_4221              Commissioner streetIlasan Lekki Lagos    600  1 beds   
lagos_4334           Ifako GbagadaIfako Gbagada Gbagada Lagos      1  3 beds   
lagos_4510                    TransformerOkun Ajah Ajah Lagos   4000  1 beds   
abuja_295                              GwarinpaGwarinpa Abuja   1300  1 beds   
abuja_495          BY TIMBER SHADE, BWARI, Central Area Abuja  90000  1 beds   
abuja_631                                      ...Lugbe Abuja    600  1 beds   
abuja_688               Wuse Zone 7 Harper StreetWuse 2 Abuja  90000  6 beds   
oyo_304                                            Ibadan Oyo  80000  1 beds   
oyo_449     U.i-Ojoo old bank road expressway IbadanIbadan...  80000  1 beds   
oyo_462      Alawaye street off akanran roadIwo Rd Ibadan Oyo  75009  1 beds   
oyo_501     Alakia old ife road inside airport second gate...  60000  1 beds   
ogun_18     Itele ogun state close to ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_31     Itele Ogun state close to Ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_32     Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
ogun_36     Itele Ogun State close to Ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_47     Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
rivers_94   Golf Estate, end of Peter Odili RoadTrans Amad...  35000  2 beds   

              baths    toilets  \
city_id                          
lagos_178   4 baths  4 Toilets   
lagos_204   1 baths  1 Toi

While the apartments in Ogun and Rivers might be truly cheap, those in Lagos and Abuja might be due to data inconsistencies.

In [109]:
exempted_towns = ['korodu', 'ogun', 'Ayobo', 'Ibadan']
for town in exempted_towns:
    cheap_apartments = cheap_apartments[~cheap_apartments.neighborhood.str.contains(town)]
cheap_apartments

neighborhood  price    beds  \
city_id                                                                        
lagos_178                           Lekki Phase 1 Lekki Lagos  55000  3 beds   
lagos_204                             ...Abule Oja Yaba Lagos  65000  1 beds   
lagos_397   Shonibare Estate Mobolaji Bank Anthony Way Ike...  35000  1 beds   
lagos_430                         Parkview Estate Ikoyi Lagos  70000  3 beds   
lagos_802   Madonna Estate ojodu Abiodun road Berger Ojodu...    250  1 beds   
lagos_1344                                   Akoka Yaba Lagos  10000    beds   
lagos_1753                           Ojodu Berger Ojodu Lagos    500  3 beds   
lagos_1866                              Phase 2 Gbagada Lagos  25000  1 beds   
lagos_2136  Off freedom road lekki phase 1Lekki Phase 1 Le...  40000  1 beds   
lagos_2213  Off studio 24, Lekki phase 1 Lekki Phase 1 Lek...  45000  2 beds   
lagos_2493             Rumens Road IkoyiOld Ikoyi Ikoyi Lagos      6  1 beds   
lagos_2543                Off Lekki Epe Expressway Ajah Lagos  70000  2 beds   
lagos_2615  Very decent and beautiful new house 2bedroom a...  50000  2 beds   
lagos_2697                         Spar roadIkate Lekki Lagos  45000  1 beds   
lagos_2839  Asekun street off adekunle kuyeAdelabu Suruler...    500  1 beds   
lagos_3112  Off Admirathy road Lekki phase 1 Lekki Phase 1...  25000  1 beds   
lagos_3208  Okotie Eboh street off Awolowo road ikoyiAwolo...  40000  1 beds   
lagos_3222                 College road Ifako Ogba Ogba Lagos    270    beds   
lagos_3225        3 Greenville estate badoreBadore Ajah Lagos    150  1 beds   
lagos_3241          Abule-ijesha road Abule Ijesha Yaba Lagos    350    beds   
lagos_3320                          Lekki Phase 1 Lekki Lagos  50000  3 beds   
lagos_4221              Commissioner streetIlasan Lekki Lagos    600  1 beds   
lagos_4334           Ifako GbagadaIfako Gbagada Gbagada Lagos      1  3 beds   
lagos_4510                    TransformerOkun Ajah Ajah Lagos   4000  1 beds   
abuja_295                              GwarinpaGwarinpa Abuja   1300  1 beds   
abuja_495          BY TIMBER SHADE, BWARI, Central Area Abuja  90000  1 beds   
abuja_631                                      ...Lugbe Abuja    600  1 beds   
abuja_688               Wuse Zone 7 Harper StreetWuse 2 Abuja  90000  6 beds   
rivers_94   Golf Estate, end of Peter Odili RoadTrans Amad...  35000  2 beds   

              baths    toilets  \
city_id                          
lagos_178   4 baths  4 Toilets   
lagos_204   1 baths  1 Toilets   
lagos_397   1 baths  2 Toilets   
lagos_430   3 baths  4 Toilets   
lagos_802   1 baths  1 Toilets   
lagos_1344    baths    Toilets   
lagos_1753  3 baths  2 Toilets   
lagos_1866  1 baths  1 Toilets   
lagos_2136  1 baths  2 Toilets   
lagos_2213  1 baths  2 Toilets   
lagos_2493  2 baths  2 Toilets   
lagos_2543  2 baths  3 Toilets   
lagos_2615  2 baths  2 Toilets   
lagos_2697  1 baths  2 Toilets   
lagos_2839  1 baths  1 Toilets   
lagos_3112  1 baths  1 Toilets   
lagos_3208  1 baths  2 Toilets   
lagos_3222    baths    Toilets   
lagos_3225  1 baths  1 Toilets   
lagos_3241    baths    Toilets   
lagos_3320  3 baths  4 Toilets   
lagos_4221  1 baths  1 Toilets   
lagos_4334  3 baths  3 Toilets   
lagos_4510  1 baths  1 Toilets   
abuja_295   1 baths  2 Toilets   
abuja_495   0 baths  0 Toilets   
abuja_631   2 baths  2 Toilets   
abuja_688   2 baths  2 Toilets   
rivers_94   2 baths  2 Toilets   

                                                        extra  
city_id                                                        
lagos_178   Furnished Serviced Newly BuiltThis is a shorle...  
lagos_204   ServicedSERVICED & SPACIOUS ROOMSELF FOR RENT....  
lagos_397   Furnished ServicedThe property is one bedroom ...  
lagos_430   Furnished3 bedroom Flat / Apartment for rent P...  
lagos_802   FurnishedA room self contain for rent at madon...  
lagos_1344  HOT SHORT-LET ?????? serviced Studio Apartment...  
l

Now we can drop the truly ridiculous ones;

In [110]:
data.drop(cheap_apartments.index, inplace=True)
data.head()

neighborhood    price    beds  \
city_id                                                                       
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  2500000  3 beds   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  2500000  3 beds   
lagos_3                       ...Lekki Phase 1 Lekki Lagos   800000  1 beds   
lagos_4                          Lekki Phase 1 Lekki Lagos  1200000  1 beds   
lagos_5                                         Ketu Lagos  1000000  2 beds   

           baths    toilets                                              extra  
city_id                                                                         
lagos_0  4 baths  4 Toilets  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  3 baths  4 Toilets  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_3  1 baths  1 Toilets  For more information, Pricing, and Private Vie...  
lagos_4  1 baths  1 Toilets  For more information, Pricing, and Private Vie...  
lagos_5  2 baths  2 Toilets  Brand new 2 bedrooms flat for rent at Ketu Lag...

### Cleaning Beds, Baths, Toilets;

In [111]:
data.beds.unique()

array(['3 beds', '1 beds', '2 beds', 'beds', '4 beds', '0 beds', '5 beds',
       '6 beds', '3 beds ', '10 beds', 'beds ', '8 beds'], dtype=object)

Some beds don't have a digit value, while some have 0.

In [112]:
no_beds = data[data.beds.str.startswith('beds')]
no_beds

neighborhood    price  beds   baths    toilets  \
city_id                                                                         
lagos_7        Salem LekkiIkate Lekki Lagos  1000000  beds   baths    Toilets   
lagos_12           Akinola Akoka Yaba Lagos   400000  beds   baths  1 Toilets   
lagos_17        St.Finbars Akoka Yaba Lagos   500000  beds   baths  1 Toilets   
lagos_18       St. Finbars Akoka Yaba Lagos   600000  beds   baths  1 Toilets   
lagos_31            Lawanson Surulere Lagos   300000  beds   baths    Toilets   
...                                     ...      ...   ...     ...        ...   
rivers_84  Trans Amadi Port Harcourt Rivers   550000  beds   baths    Toilets   
rivers_85  Trans Amadi Port Harcourt Rivers   600000  beds   baths    Toilets   
rivers_86  Trans Amadi Port Harcourt Rivers   800000  beds   baths    Toilets   
rivers_88       Eliozu Port Harcourt Rivers   800000  beds   baths    Toilets   
rivers_90  Trans Amadi Port Harcourt Rivers   400000  beds   baths    Toilets   

                                                       extra  
city_id                                                       
lagos_7    Nice and Standard Renovated Mini Flat availabl...  
lagos_12   Furnished ServicedThis is a lovely SelfContain...  
lagos_17   Furnished Serviced Newly BuiltThis is a lovely...  
lagos_18   Furnished Serviced Newly BuiltThis is a lovely...  
lagos_31   *Clean Miniflat with tiles and water at idiara...  
...                                                      ...  
rivers_84  Cool 2 Bedroom Flat For Rent in Woji Port Harc...  
rivers_85  Modern King Size 2 Bedroom Flat For Rent In Wo...  
rivers_86  Tastefully Finished 2 Bedroom Flat For Rent in...  
rivers_88  Well Finished Brand new 2 Bedroom Flat For Ren...  
rivers_90  Decent Size 2 Bedroom Flat For Rent In Woji Po...  

[349 rows x 6 columns]

In [113]:
data.drop(no_beds.index, inplace=True)

data.beds.unique()

array(['3 beds', '1 beds', '2 beds', '4 beds', '0 beds', '5 beds',
       '6 beds', '3 beds ', '10 beds', '8 beds'], dtype=object)

In [114]:
data.beds = data.beds.map(lambda x: x[0]).astype(int)

data.beds.unique()

array([3, 1, 2, 4, 0, 5, 6, 8])

Great! Now onto the 0 beds;

In [115]:
data[data.beds == 0]

neighborhood    price  beds  \
city_id                                                                       
lagos_65   Close to 3rd Mainland Bridge, Yaba, Lagos. Yab...   800000     0   
lagos_212                       ...Adeniyi Jones Ikeja Lagos  2000000     0   
lagos_254                           Chemist Akoka Yaba Lagos   400000     0   
lagos_268         Obadia- off ilaje roadBariga Shomolu Lagos   350000     0   
lagos_441     destiny home estate Abijo AJAHAbijo Ajah Lagos   500000     0   
...                                                      ...      ...   ...   
rivers_61                ...Trans Amadi Port Harcourt Rivers   800000     0   
rivers_75                 - Trans Amadi Port Harcourt Rivers   750000     0   
rivers_82                ...Trans Amadi Port Harcourt Rivers   500000     0   
rivers_83                 - Trans Amadi Port Harcourt Rivers  1000000     0   
rivers_89                  -Trans Amadi Port Harcourt Rivers  1200000     0   

             baths    toilets  \
city_id                         
lagos_65   0 baths  0 Toilets   
lagos_212  2 baths  3 Toilets   
lagos_254  0 baths  1 Toilets   
lagos_268  0 baths  0 Toilets   
lagos_441  0 baths  0 Toilets   
...            ...        ...   
rivers_61  0 baths  0 Toilets   
rivers_75  0 baths  0 Toilets   
rivers_82  0 baths  0 Toilets   
rivers_83  0 baths  0 Toilets   
rivers_89  0 baths  0 Toilets   

                                                       extra  
city_id                                                       
lagos_65   Currently Being Renovated: Spacious Miniflat (...  
lagos_212  JUST OUT IN ADENIYI JONES 2 bedroom Luxery apa...  
lagos_254  ServicedThis is a lovely SelfContain @ Akoka Y...  
lagos_268  FurnishedA Newly Renovated Mini Flat @ Obadiah...  
lagos_441  Room and parlour self contain for rent at dest...  
...                                                      ...  
rivers_61  Modern King Size 3 Bedroom Flat For Rent In Wo...  
rivers_75  Beautiful 1 Bedroom Flat For Rent In Woji Port...  
rivers_82  Neat 2 Bedroom Flat For Rent in Woji Port Harc...  
rivers_83  Decent king size 2 Bedroom Flat For Rent In Pe...  
rivers_89  Decent 2 Bedroom Flat For Rent in Peter Odili ...  

[161 rows x 6 columns]

Some of them are just plain errors and are already in the description column.

In [118]:
for i in range(1, 6):
    data.loc[data.extra.str.contains(str(i)), 'beds'] = i
data.loc[data.extra.str.contains('two'), 'beds'] = 2 # for those written in words
data.loc[data.extra.str.contains('three'), 'beds'] = 3 # for those written in words

data.loc[data.extra.str.contains('and parlour'), 'beds'] = 1 # for 'room and parlour' apartments
data.loc[data.extra.str.contains('elf'), 'beds'] = 1 # for 'self contained' apartments

bed_zero = data[data.beds == 0]
bed_zero

neighborhood    price  beds  \
city_id                                                                        
lagos_467                               ...Ilasan Lekki Lagos  1000000     0   
lagos_500                            Harmony estateAjah Lagos   400000     0   
lagos_656           Nice and Secure EstateChevron Lekki Lagos  1500000     0   
lagos_1683  Abule oja, second gate, yaba-lagos..Abule Oja ...   400000     0   
lagos_1706           Green Field EstateAgo Palace Okota Lagos   700000     0   
lagos_1980               Arowojobe EstateMende Maryland Lagos  1000000     0   
lagos_2118                  Bama villaAbule Ijesha Yaba Lagos   400000     0   
lagos_2310                        SomodiShomolu Shomolu Lagos   500000     0   
lagos_2624  Idaosa ojo osagie, off peace estate AgudaAguda...   700000     0   
lagos_2849                          ...Ago Palace Okota Lagos   450000     0   
lagos_2902                           ...Ikeja Gra Ikeja Lagos  2000000     0   
lagos_3343              Amule bus stop AyoboAyobo Ipaja Lagos   200000     0   
lagos_3387    nnobi street,kilo, surulere,LagosSurulere Lagos   650000     0   
lagos_3455                          BOGIJE, ELEMOROAjah Lagos   350000     0   
lagos_3985                                  AlapereKetu Lagos   450000     0   
lagos_4053  ShopRite Road Osapa London LekkiOsapa London L...   350000     0   
lagos_4250        S Forte Estate, Before BlencoAdo Ajah Lagos   650000     0   
lagos_4325                    Iju ishagaIfako Ogba Ogba Lagos   300000     0   
lagos_4473                          ...Ebute Metta Yaba Lagos   350000     0   
oyo_431          First Power line Eleyele/OloguneruIbadan Oyo   150000     0   
oyo_433                           Old BodijaBodija Ibadan Oyo   400000     0   

               baths     toilets  \
city_id                            
lagos_467    0 baths   0 Toilets   
lagos_500    0 baths   0 Toilets   
lagos_656    0 baths   0 Toilets   
lagos_1683   0 baths   0 Toilets   
lagos_1706   0 baths   0 Toilets   
lagos_1980   0 baths   0 Toilets   
lagos_2118   1 baths   1 Toilets   
lagos_2310   0 baths   0 Toilets   
lagos_2624   0 baths   0 Toilets   
lagos_2849   0 baths   0 Toilets   
lagos_2902   0 baths   0 Toilets   
lagos_3343   0 baths   0 Toilets   
lagos_3387   0 baths   0 Toilets   
lagos_3455   0 baths   0 Toilets   
lagos_3985   0 baths   0 Toilets   
lagos_4053   0 baths   0 Toilets   
lagos_4250   0 baths   0 Toilets   
lagos_4325   0 baths   0 Toilets   
lagos_4473   0 baths   0 Toilets   
oyo_431      0 baths   0 Toilets   
oyo_433      0 baths   0 Toilets   

                                                        extra  
city_id                                                        
lagos_467   JUST OUT!! SEMI DIRECT!! VERY CLEAN MINI FLAT ...  
lagos_500               Very spacious and neat environment..   
lagos_656   ServicedNice and Standard Mini Flat with inten...  
lagos_1683  FurnishedVery specious and standard mini flat ...  
lagos_1706  Newly built Executive Lovely Sharp Mini Flat o...  
lagos_1980  Brand new Tastefully finished miniflat with en...  
lagos_2118  FurnishedHOT LETTING: Lovely built serviced ap...  
lagos_2310  Mini flat Flat / Apartment for rent Somodi Sho...  
lagos_2624  FurnishedTOLET A Tastefully finished,spacious ...  
lagos_2849  FurnishedStandard Mini Flat on a good Environm...  
lagos_2902  Brand new, exquisitely finished and fully furn...  
lagos_3343  A newly built mini flat at the back of access ...  
lagos_3387  For lease: Fully finished mini flat with wardr...  
lagos_3455  Tastefully finished POP mini flat, spacious, w...  
lagos_3985  Newly BuiltNewly built pop finished , spacious...  
lagos_4053  Serviced Newly BuiltNice and Standard Newly Bu...  
lagos_4250  Flat / Apartment for rent S Forte Estate, Befo...  
lagos_4325  *BRAND NEW EXECUTIVE MINI FLAT AT IJU ISHAGA* ...  
lagos_4473  VERY SPACIOUS & DECENT NEWLY RENOVATED ROOM SE...  
oyo_431     It is located in a well

And I've reduced 161 unknown rooms to just a handful!

In [120]:
data.drop(bed_zero.index, inplace=True)

data.beds.unique()

array([3, 1, 2, 5, 4])

**Moving on to Baths;**

In [121]:
data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', ' baths', '0 baths',
       ' 2 baths', '5 baths', ' 3 baths', ' 1 baths', '  baths',
       '6 baths', '10 baths', ' 5 baths'], dtype=object)

In [122]:
data.baths = data.baths.str.strip()

data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', 'baths', '0 baths',
       '5 baths', '6 baths', '10 baths'], dtype=object)

In [124]:
no_baths = data[data.baths.str.startswith('baths')]
no_baths

neighborhood    price  beds  \
city_id                                                                       
lagos_26                                         Agege Lagos   800000     3   
lagos_28            Off Demurin road Ketu Kosofe Ikosi Lagos   400000     1   
lagos_29            Off Demurin road Ketu Kosofe Ikosi Lagos  1000000     2   
lagos_30              Seaside estate badoreBadore Ajah Lagos  1800000     4   
lagos_68   Alagomeji-Adekunle, Close to the 3rd Mainland ...  1500000     2   
...                                                      ...      ...   ...   
ogun_108                                  Obafemi Owode Ogun   400000     2   
ogun_153                Arepo private EstateArepo Arepo Ogun   500000     2   
ogun_182   Agbeloba upright hotel street Abalawa Abeokuta...   250000     1   
rivers_81                   Trans Amadi Port Harcourt Rivers   400000     1   
rivers_87                        Eliozu Port Harcourt Rivers  1000000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_26   baths    Toilets  Brand new 3 bedroom flat in Oko Oba, Agege for...  
lagos_28   baths    Toilets  Newly renovated upstairs roomself apartment pr...  
lagos_29   baths    Toilets  Newly renovated 2 bedrooms flats with dining, ...  
lagos_30   baths    Toilets  4bedroom duplex, alone in the compound, pop fi...  
lagos_68   baths  2 Toilets  Lovely New & Very Spacious Executive 2 Bedroom...  
...          ...        ...                                                ...  
ogun_108   baths    Toilets  FurnishedA Nice 2 Bedroom Apartment in a Priva...  
ogun_153   baths  3 Toilets  Furnished Newly BuiltLovely 2 bed room flat at...  
ogun_182   baths  1 Toilets  Newly BuiltRoom and parlour self contained wit...  
rivers_81  baths    Toilets  Tastefully Finished New 1 Bedroom Flat For Ren...  
rivers_87  baths    Toilets  Decent Brand new 2 Bedroom Flat For Rent In Sh...  

[317 rows x 6 columns]

**Dealing with Toilets;**

In [125]:
data.toilets.unique()

array(['4 Toilets', '1 Toilets', '2 Toilets', '3 Toilets', ' Toilets',
       ' 4 Toilets', '0 Toilets', ' 1 Toilets', '5 Toilets', ' 0 Toilets',
       '6 Toilets', ' 3 Toilets', ' 2 Toilets', '10 Toilets', '  Toilets'],
      dtype=object)

In [126]:
data.toilets = data.toilets.str.strip()

data.toilets.unique()

array(['4 Toilets', '1 Toilets', '2 Toilets', '3 Toilets', 'Toilets',
       '0 Toilets', '5 Toilets', '6 Toilets', '10 Toilets'], dtype=object)

In [128]:
data[data.toilets.str.startswith('T')]

neighborhood    price  beds    baths  \
city_id                                                                       
lagos_26                                Agege Lagos   800000     3    baths   
lagos_28   Off Demurin road Ketu Kosofe Ikosi Lagos   400000     1    baths   
lagos_29   Off Demurin road Ketu Kosofe Ikosi Lagos  1000000     2    baths   
lagos_30     Seaside estate badoreBadore Ajah Lagos  1800000     4    baths   
lagos_161                      Masha Surulere Lagos  1600000     3  2 baths   
...                                             ...      ...   ...      ...   
rivers_70          Trans Amadi Port Harcourt Rivers  1100000     3  3 baths   
rivers_72         -Trans Amadi Port Harcourt Rivers  1500000     3  3 baths   
rivers_76          Trans Amadi Port Harcourt Rivers  1000000     3  3 baths   
rivers_81          Trans Amadi Port Harcourt Rivers   400000     1    baths   
rivers_87               Eliozu Port Harcourt Rivers  1000000     2    baths   

           toilets                                              extra  
city_id                                                                
lagos_26   Toilets  Brand new 3 bedroom flat in Oko Oba, Agege for...  
lagos_28   Toilets  Newly renovated upstairs roomself apartment pr...  
lagos_29   Toilets  Newly renovated 2 bedrooms flats with dining, ...  
lagos_30   Toilets  4bedroom duplex, alone in the compound, pop fi...  
lagos_161  Toilets  ServicedFOR RENT... A SERVICED & EXECUTIVE 3BE...  
...            ...                                                ...  
rivers_70  Toilets  Spacious 3 Bedroom Flat For Rent in Peter Odil...  
rivers_72  Toilets  Tastefully Built 3 Bedroom Flat For Rent in Pe...  
rivers_76  Toilets  Executive Newly Built 3 Bedroom Flat For Rent ...  
rivers_81  Toilets  Tastefully Finished New 1 Bedroom Flat For Ren...  
rivers_87  Toilets  Decent Brand new 2 Bedroom Flat For Rent In Sh...  

[392 rows x 6 columns]